File: 2020/13_ShuttleSearch/bus.py
# ======================================================================
# Shuttle Search
#   Advent of Code 2020 Day 13 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                           b u s . p y
# ======================================================================
"A people mover for the Advent of Code 2020 Day 13 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                    Bus
# ======================================================================


class Bus(object):   # pylint: disable=R0902, R0205
    "Object for Shuttle Search"

    def __init__(self, bid=None, offset=0):

        # 1. Set the initial values
        self.bid = 0
        self.offset = offset

        # 2. Set bid if one is given
        if bid is not None:
            self.bid = int(bid)

    def next_depart_after(self, time):
        "Return when the next time the bus departs after specified time"
        delta = time % self.bid
        if delta == 0:
            return time
        return time + self.bid - delta


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                            b u s . p y                         end
# ======================================================================

File: 2020/13_ShuttleSearch/bus.lua
-- ======================================================================
-- Shuttle Search
--   Advent of Code 2020 Day 13 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                           b u s . l u a
-- ======================================================================
-- Bus for the Advent of Code 2020 Day 13 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Bus = { part2=false, bid=0, offset=0 }
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                                    Bus
-- ======================================================================


function Bus:Bus (o)
  -- Object for Shuttle Search


  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.bid = o.bid or 0
  o.offset = o.offset or 0

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  return o
end

function Bus:next_departure(time)
  -- Return the next time the bus departs after specified time
  time = time or 0
  
  local delta = time % self.bid
  if delta == 0 then
    return time
  end  
  return time + self.bid - delta
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
return Bus

-- ======================================================================
-- end                          b u s . l u a                         end
-- ======================================================================
